### PR TITLE
CLOUDSTACK-10380: Fix startvm giving another pw after pw reset

### PR DESCRIPTION
**Pull request by smeetsr**
## Description

* When password reset is called, and VR accepts it, set vm updateParameters to false.
* When password reset is called, and SSH key is set, make sure the password isn't removed again.
* Fix Marvin tests
* Make sure saveDetails keeps the display value of the existing details.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
 - Manual 
 - automated marvin pugins nuagevsp tests on ACS + NuageVsp
 - automated marvin component test for config drive

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [x] I have added tests to cover my changes.
- [ ] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: test/integration/component/test_configdrive.py [hunk 1]
```diff
@@ -45,21 +49,17 @@
...
+                             Hypervisor, Template)
 from marvin.lib.common import (get_domain,
                                get_template,
-                               get_zone,
-                               list_templates)
+                               get_zone)
 from marvin.lib.utils import random_gen
 # Import System Modules
 from nose.plugins.attrib import attr
 from retry import retry
```
Comment (borisstoyanov) [inline test/integration/component/test_configdrive.py]: this dependency is not in the required list in marvin installation

**Comment by borisstoyanov** [comment]
there's seems to be an issue with the vm password
I'm able to ssh using default password, looks like the tests are either using some other hard-coded one or trying to set a new and that part does not seem to work.

File: server/src/com/cloud/network/element/VirtualRouterElement.java [hunk 2]
```diff
@@ -703,7 +703,14 @@ public boolean savePassword(final Network network, final NicProfile nic, final V
         // save the password in DB
         for (final VirtualRouter router : routers) {
             if (router.getState() == State.Running) {
-                return networkTopology.savePasswordToRouter(network, nic, uservm, router);
+                final boolean result = networkTopology.savePasswordToRouter(network, nic, uservm, router);
+                if (result) {
+                    // Explicit password reset, while VM hasn't generated a password yet.
+                    final UserVmVO userVmVO = _userVmDao.findById(vm.getId());
+                    userVmVO.setUpdateParameters(false);
```
Comment (rafaelweingartner) [inline server/src/com/cloud/network/element/VirtualRouterElement.java]: What does this `updateParameters` means in the `userVmVo` object?

File: server/src/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -688,10 +689,6 @@ public UserVm resetVMPassword(ResetVMPasswordCmd cmd, String password) throws Re
 
         if (result) {
             userVm.setPassword(password);
-            // update the password in vm_details table too
-            // Check if an SSH key pair was selected for the instance and if so
-            // use it to encrypt & save the vm password
-            encryptAndStorePassword(userVm, password);
```
Comment (rafaelweingartner) [inline server/src/com/cloud/vm/UserVmManagerImpl.java]: So, the password will not be encrypted anymore in the DB? Is that it?

**Comment by smeetsr** [comment]
Be aware that Frank Maximus is on holiday, he will be back Monday July 30th. I will make sure he responds to your questions asap.
Here I am putting the Marvin runinfo and result files for component/test_configdrive.py on my testbed.
I have installed Marvin with nuagevsp exension pip install 'Marvin-4.11.1.0-SNAPSHOT.tar.gz[nuagevsp]' maybe this is causing the issue upstream. Indeed the standard password of our guest vm template is password.
[results_of_test_configdrive_isolated_network.txt](https://github.com/apache/cloudstack/files/2210343/results_of_test_configdrive_isolated_network.txt)
[runinfo_of_test_configdrive_isolated_network.txt](https://github.com/apache/cloudstack/files/2210344/runinfo_of_test_configdrive_isolated_network.txt)
[results_of_test_configdrive_vpc_network.txt](https://github.com/apache/cloudstack/files/2210351/results_of_test_configdrive_vpc_network.txt)
[runinfo_of_test_configdrive_vpc_network.txt](https://github.com/apache/cloudstack/files/2210350/runinfo_of_test_configdrive_vpc_network.txt)
I also just pushed a small fix in plugins/nuagevsp/test_nuage_extra_dhcp.py

File: test/integration/component/test_configdrive.py (hunk 1 above)
Comment (fmaximus) [reply test/integration/component/test_configdrive.py]: `retries` was added as as install dependency in tools/marvin/setup.py

File: server/src/com/cloud/vm/UserVmManagerImpl.java (hunk 3 above)
Comment (fmaximus) [reply server/src/com/cloud/vm/UserVmManagerImpl.java]: encryptAndStorePassword was already done inside resetVMPasswordInternal.

File: server/src/com/cloud/network/element/VirtualRouterElement.java (hunk 2 above)
Comment (fmaximus) [reply server/src/com/cloud/network/element/VirtualRouterElement.java]: Context: Basically `updateParameters` means that a new password needs to be generated when the VM starts.
This is set  to `true` when resetPassword is called, while there is no VR running.
Also the default is `true` for a new VM. When the template was not password enabled during VM creation, 
this stays `true`. So if `resetPassword` is executed after making the template password enabled, and VR is running,
we set if to `false`, as we successfully did a password reset, thus no password reset needs to happen during start VM.

**Comment by fmaximus** [comment]
@borisstoyanov The Marvin test failures you mention seem to refer to an older version of `test_configdrive`. Also even though the import says `retry` the library is named `retries` (there are about 5 retry libraries around).

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by fmaximus** [comment]
@borisstoyanov We are using an updated version of the built-in guest template, which includes the `cloud-set-guest-password-configdrive` script, besides `cloud-set-guest-password`.

**Comment by borisstoyanov** [comment]
@fmaximus as a user I would expect to be able to run the tests in the community repo with the stock installation (including templates). I can suggest you to upload the template somewhere public (http://dl.openvm.eu/cloudstack/ for example) and use it in the test if it requires specific scripts to be in it.

**Comment by fmaximus** [comment]
@borisstoyanov I've put the template here: http://cs.mv.nuagenetworks.net/images/centos55-extended.qcow2.xz

What I get from your comment, is that you would like to have the test register the template in the test.
This will not be an easy task for us, as we're behind a proxy, and I haven't been able to get ssvm go through the proxy yet (by using the global settings).

But in the end the default test template should be able to handle config drive, 
for which there are 2 options:
1. use a cloud-init enabled centos 7 default template. (preferred, already being discussed)
2. add the scripts to the existing built-in template

**Comment by borisstoyanov** [comment]
Hi @fmaximus, we've got similar issues with our trillian envs and in order to make download faster we've setup mirrors locally so we could download it really fast. You could use this same approach and host the template to local space where your SSVMs could access. That would of course require to hot-fix the test with different urls post-installation. 
I think a publicly hosted template with all required scripts would work for community, and if people would have issues accessing the resource will be only due to their own setup.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
This is a fairly large PR, I'll review it soon but exclude reviewing nuage/cfgdrive specific marvin tests and changes that I won't be able to run/validate.

**Comment by rhtyd** [comment]
@smeetsr @fmaximus can you share your test results from nuage specific marvin tests?
@borisstoyanov can you help validate regression tests for config drive related features?
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@fmaximus, @smeetsr are you guys planning to work on the template that would enable the execution of the config drive tests, rest of the smoketests looks good as far as I see.

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Test-wise it LGTM, @borisstoyanov are you okay or have found regressions with this PR? Can @smeetsr @fmaximus comment on raised issues?

**Comment by rhtyd** [comment]
Please comment @smeetsr @fmaximus for @borisstoyanov 's review

**Comment by fmaximus** [comment]
@borisstoyanov @rhtyd 
We're currently working on changing the tests to register the template.

**Comment by PaulAngus** [comment]
Hi @smeetsr @fmaximus how are the tests coming along? we'd really like to get this merged ASAP.

**Comment by fmaximus** [comment]
Component test is updated to register the extended CentOS 5.5 image. Also got proxy on secondary storage working. We would like to extend macchinina to support config drive though, so the tests can run faster.

**Comment by PaulAngus** [comment]
@blueorangutan package

**Comment by PaulAngus** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by fmaximus** [comment]
@borisstoyanov, are you ok with the latest changes?

**Comment by fmaximus** [comment]
@borisstoyanov @PaulAngus That seems to be an issue in Cloudstack, not in the test.
The VM doesn't seem to have the config drive attached.

**Comment by smeetsr** [comment]
Here are the results of component/test_config_drive.py on my nativeACS setup with this PR.
[results.txt](https://github.com/apache/cloudstack/files/2401185/results.txt)
[runinfo.txt](https://github.com/apache/cloudstack/files/2401187/runinfo.txt)

**Comment by smeetsr** [comment]
My setup is centos7 based 2 KVM hypervisors with Advanced VLAN networking, nfs storage.

**Comment by borisstoyanov** [comment]
I guess this needs further investigation, as it's showing a problem in config drive (either test or code). I don't find anything related to different setup between both the test environments... :(

**Comment by DaanHoogland** [comment]
you disapproved this @PaulAngus , so there is a reason this must not be merged? or did you mean to request for changes?

**Comment by PaulAngus** [comment]
I thought that I had requested changes @DaanHoogland :(
where does it say that i disapproved?

**Comment by DaanHoogland** [comment]
sorry @PaulAngus my bad. I deducted wrongly from the red cross in combination with your text not being posed as a question.
@smeetsr can you look into @PaulAngus ' report, tnx?

**Comment by PaulAngus** [comment]
@fmaximus - could you give us a timescale on fixes for the test and the adding NIC bug?

**Comment by smeetsr** [comment]
@PaulAngus  According to me, @fmaximus  already pushed the changes to the test last Friday. The Jenkins job however failed due to out of diskspace. I guess this was an environmental issue?

**Comment by PaulAngus** [comment]
@blueorangutan package

**Comment by PaulAngus** [comment]
Hi @smeetsr @fmaximus,
The test now runs without throwing exceptions - the issue with configdrive not working when an additional NIC is added remains though.

**Comment by fmaximus** [comment]
@PaulAngus Can we decouple that issue from this bugfix? I'm in favour of creating a new ticket for the add nic issue, and merging this one.

**Comment by krissterckx** [comment]
Agreee with @fmaximus , let's separate that out ; that must be a new issue, individually tracked.
@rhtyd agree ?

**Comment by PaulAngus** [comment]
sounds sensible to me.  @rhtyd ?

**Comment by rhtyd** [comment]
+1 please send a PR separately @fmaximus @krissterckx

**Pull request merged**